### PR TITLE
Feature/optim

### PR DESCRIPTION
**Pull request by wang-chen**
Add Gauss-Newton Optimizer and Robust Kernel function such as Huber.

### CODE REVIEW

File: pypose/optim/optimizer.py [hunk 1]
```diff
@@ -1,39 +1,244 @@
...
+            return self.model(*inputs)
+        else:
+            return self.model(inputs)
+
+    def residual(self, outputs, targets):
+        return outputs if targets is None else outputs - targets
+
+    def kernel_forward(self, module, inputs, outputs):
+        # eps is to prevent grad of sqrt() from being inf
+        eps = finfo(outputs.dtype).eps
```
Comment (huyaoyu) [inline pypose/optim/optimizer.py:49]: This raises an exception if outputs.dtype is an integer type.

File: pypose/optim/optimizer.py [hunk 2]
```diff
@@ -1,39 +1,244 @@
...
+            Early Stopping with error: 5.21540641784668e-07
+        '''
+        R = self.model(inputs, targets)
+        for pg in self.param_groups:
+            numels = [p.numel() for p in pg['params'] if p.requires_grad]
+            J = modjac(self.model, inputs=(inputs, targets), flatten=True)
+            R, J = self.corrector(R = R, J = J)
+            D = self.solver(A = J, b = -R.view(-1, 1)).split(numels)
+            [p.add_(d.view(p.shape)) for p, d in zip(pg['params'], D) if p.requires_grad]
+        return self.model.loss(inputs, targets)
```
Comment (huyaoyu) [inline pypose/optim/optimizer.py:209]: self.model is required to have the loss() function. So it is not a simple nn.module.
The same comment for the LM.

Comment (wang-chen) [reply pypose/optim/optimizer.py:209]: the loss() function is from the standardized model, not the model that user provide any more: see this line

https://github.com/pypose/pypose/blob/cca97237eb0120e31d8acf50097b70a8a400a0ef/pypose/optim/optimizer.py#L141

Comment (wang-chen) [reply pypose/optim/optimizer.py:209]: I could change the `loss` method to `_loss` if you think this name may conflict with the users' self-defined method.
